Update noops-sql-db.md (#786)
fermyon/developer: content/cloud/noops-sql-db.md

title = "NoOps SQLite Database"
template = "cloud_main"
date = "2023-07-26T16:00:00Z"
enable_shortcodes = true
[extra]
url = "https://github.com/fermyon/developer/blob/main/content/cloud/noops-sql-db.md"

---

- [Accessing Private Beta](#accessing-private-beta)
- [Quotas And Service Limitations For the Database in Fermyon Cloud](#quotas-and-service-limitations-for-the-database-in-fermyon-cloud)
- [How To](#how-to)

>   - [Database Access From Many Components](#database-access-from-many-components)

Reviewer: The grammar of the "how to" subtitles doesn't flow from the top-level title - "How to database access", "How to deleting", etc.

Author: Thanks @itowlson 
I am updating this in the next push to align the language for each heading.

  - [Preparing Fermyon Cloud Schema](#preparing-fermyon-cloud-schema)
  - [Deleting the Database](#deleting-the-database)
- [FAQ](#faq)
- [Next Steps](#next-steps)

> With NoOps SQL Database using SQLite support in Fermyon Cloud, you can persist relational data generated by your [Spin](../spin/install.md) application in between application invocations without worrying about database management. Fermyon Cloud will provision and manage the database on your behalf; no operations (NoOps) are required. Fermyon Cloud will do the heavy lifting for you. To learn more about the Spin SQLite SDK, please visit the [API guide](../spin/sqlite-api-guide.md).

Reviewer: style nit: for me the change to "we do X., no operations required" to "we do X; no operations are required" is less punchy.

Author: Thanks, I reverted in next push.

> title = "NoOps SQLite Database"

Reviewer 2: Checked and we are using "NoOps SQL" as the general name for the feature. We need to update the CLI too
```suggestion
title = "NoOps SQL Database"
```

* Your NoOps SQL Database can hold up 1 GB of data

*Service Limitations*
* The `--sqlite` flag is not supported on `spin cloud deploy`
* Fermyon **Cloud** only supports databases named "default" at this time

## How To

> Firstly, if you have not already done so, you will need to [create your first Spin application](https://developer.fermyon.com/spin/quickstart#create-your-first-application). Our blog posts [Announcing Fermyon Cloud’s NoOps SQL Database](https://www.fermyon.com/blog/announcing-noops-sql-db) and [NoOps and Serverless Are the Perfect Pair](https://www.fermyon.com/blog/noops-and-serverless-are-the-perfect-pair) each have an example of how to create a NoOps SQLite Storage application. To tell Spin that we want to use SQLite storage, we only need to [grant SQLite permission to a component in the application’s manifest]((https://developer.fermyon.com/spin/sqlite-api-guide#granting-sqlite-database-permissions-to-components)) (the spin.toml file). For example:

Reviewer: Do we really need to start from "first create your Spin application"?  To me this takes us away from "here's how to get your SQLite app running in Cloud" to "back to the Getting Started guide again."

Author: I have created a separate section for this (making it clear that creating a new application is optional) in case the reader already has an application and they just want to get it running SQLite in Cloud.

> Firstly, if you have not already done so, you will need to [create your first Spin application](https://developer.fermyon.com/spin/quickstart#create-your-first-application). Our blog posts [Announcing Fermyon Cloud’s NoOps SQL Database](https://www.fermyon.com/blog/announcing-noops-sql-db) and [NoOps and Serverless Are the Perfect Pair](https://www.fermyon.com/blog/noops-and-serverless-are-the-perfect-pair) each have an example of how to create a NoOps SQLite Storage application. To tell Spin that we want to use SQLite storage, we only need to [grant SQLite permission to a component in the application’s manifest]((https://developer.fermyon.com/spin/sqlite-api-guide#granting-sqlite-database-permissions-to-components)) (the spin.toml file). For example:

Reviewer: ```suggestion
Firstly, if you have not already done so, you will need to [create your first Spin application](/spin/quickstart#create-your-first-application). Our blog posts [Announcing Fermyon Cloud’s NoOps SQL Database](https://www.fermyon.com/blog/announcing-noops-sql-db) and [NoOps and Serverless Are the Perfect Pair](https://www.fermyon.com/blog/noops-and-serverless-are-the-perfect-pair) each have an example of how to create a NoOps SQLite Storage application. To tell Spin that we want to use SQLite storage, we only need to [grant SQLite permission to a component in the application’s manifest](/spin/sqlite-api-guide#granting-sqlite-database-permissions-to-components) (the spin.toml file). For example:
```

Author: Good catch @itowlson 
I have updated all links to developer subdomain to be relative.

```
[component]
sqlite_databases = ["default"]
```

> ### Database Access From Many Components

Reviewer: I'm a bit puzzled by this title.  The original intent of this section was to explain that the database was created at deployment time and persisted across deployments.  I guess the "how to give access" stuff confused the issue but that was meant to be incidental.  I am not sure what this section is now intending to show.

Author: Thanks, this has been updated to suite Cloud private preview context.

> A database is created when you build a Spin application referencing a "default" SQLite database in the component section of an application's manifest. Multiple components in the same Spin application can be given access to the same `default` database like so:

Reviewer: "deploy" was correct - `spin build` does _not_ create a database.  (Also I prefer "that references" because it can only apply to "application", whereas "referencing" could modify "build".)

```suggestion
A database is created when you deploy a Spin application that references a "default" SQLite database in the component section of an application's manifest. Multiple components in the same Spin application can be given access to the same `default` database like so:
```

Author: Ah, yes. Apologies. I had it in my head that `build` created the database environment due to my following workflow example (below). I just now realized that the `--sqlite` I am using is an option of an `up` subcommand. So technically, the database is indeed being created and populated as part of a deployment, not a `build`. Thanks for pointing this out.

```
$ spin new http-rust
Enter a name for your new application: test_sqlite
Description: Test SQLite
HTTP base: /
HTTP path: /...
$ cd test_sqlite 
# add the sqlite_databases = ["default"] to manifest file
$ spin build
$ spin up --sqlite "CREATE TABLE IF NOT EXISTS character (id INTEGER PRIMARY KEY AUTOINCREMENT, firstname TEXT NOT NULL, lastname TEXT NOT NULL)" --sqlite "INSERT INTO character VALUES (NULL,'Lois','Lane')"
$ spin cloud sqlite list
Databases (1)
vivacious-mangosteen (default)
```

Author: I will commit this as part of my next push; thanks for the wording.

Reviewer: Yes, but regardless, this is _NOT_ creating the cloud database.  `spin cloud sqlite list` is listing databases that exist in the cloud (which are created on `deploy`.  `spin up` creates (if needed) a _local_ database, by default in the application's `.spin` directory.  Your mangosteen database is one that happened to already exist in your cloud account.

> A database is created when you build a Spin application referencing a "default" SQLite database in the component section of an application's manifest. Multiple components in the same Spin application can be given access to the same `default` database like so:

Reviewer: The more I look at this, the more I think we should drop the stuff about "multiple components" - it just recapitulates one of the less interesting parts of the Spin docs, and points readers in the wrong direction.  We could show _one_ component if we really wanted to but even then I kind of feel we should decouple thing from the "here is how the database gets created" stuff.

Author: Agreed, this has been dropped.

* Your NoOps SQL Database can hold up 1 GB of data

*Service Limitations*
* The `--sqlite` flag is not supported on `spin cloud deploy`
* Fermyon **Cloud** only supports databases named "default" at this time

## Creating a New Spin Application (Optional)

If you already have a Spin application, you can skip this step. If you do not have a Spin application on localhost you will need to [create one](/spin/quickstart#create-your-first-application).

## Grant SQLite Permission

To tell Spin that we want to use SQLite storage, we only need to [grant SQLite permission to a component in the application’s manifest]((/spin/sqlite-api-guide#granting-sqlite-database-permissions-to-components)) (the spin.toml file). For example:

```
[component]
sqlite_databases = ["default"]
```

The new database file (`sqlite_db.db`) is created in your application's `.spin` folder when you deploy this Spin application. The database persists Spin application invocations and updates. The database is empty and can optionally have tables and data added using the following step.

> ## Tables and Data Locally (Optional)

Author: Thanks, I have removed the `optional` from the heading but will leave the instructions because running this locally is worth doing so users can test out how their application code interacts with the database before deploying to the cloud.

## Tables and Data in Fermyon Cloud

When you first `spin cloud deploy` your application (with the `sqlite_databases` configuration, as shown above), Fermyon Cloud will create a new & empty database for your Cloud application. This new Cloud database will have a unique name. To obtain the name, type the following command in your local NoOps SQLite application's directory:

<!-- @selectiveCpy -->

```bash
$ spin cloud sqlite list
Databases (1)
inspirational-pig (default)
```

Then use the `spin cloud sqlite execute` command to create your Cloud database's schema i.e. execute language syntax compatible with [SQLite](https://www.sqlite.org/lang.html) against your Cloud database.

```bash
$ spin cloud sqlite execute inspirational-pig "CREATE TABLE IF NOT EXISTS todos (id INTEGER PRIMARY KEY AUTOINCREMENT,description TEXT NOT NULL,due_date DATE,starred BOOLEAN DEFAULT 0,is_completed BOOLEAN DEFAULT 0)"
```

> Note: Cloud database is completely unrelated to the local (optional) database.

> ## Deleting the Cloud Database

Author: After Kate's suggestions and the last round of changes, this is implied now.

Warning: using the `delete` subcommand is **permanent**. 

You can delete your default database with the following command:

```bash
$ spin cloud sqlite delete inspirational-pig
The action is irreversible.
Please type "inspirational-pig" for confirmation: inspirational-pig
Deleting database ...
Database "inspirational-pig" deleted
```

> > In the Private Beta, the database is inaccessible once the application has been deleted. However, we separate out the database deletion step because we anticipate the database being accessible after deletion in a future beta. We enforce database deletion to prevent unintentional data sharing between Spin applications. 

Reviewer 2: The database is currently accessible after an application is deleted. It is that we currently do not support moving a database from one app to another.

Another point to this is that we do not currently support exporting a database's data before deletion. I think that is a goal too.
```suggestion
> In the Private Beta, we do not have support for sharing database between apps. However, we still separate out the database deletion from app deletion because we anticipate the ability to use databases from multiple apps in the future.
```

* Your NoOps SQL Database can hold up 1 GB of data

*Service Limitations*
* The `--sqlite` flag is not supported on `spin cloud deploy`
* Fermyon **Cloud** only supports databases named "default" at this time

## Creating a New Spin Application

If you already have a Spin application, you can skip this step. If you do not have a Spin application locally you will need to [create one](/spin/quickstart#create-your-first-application).

## Grant SQLite Permission

To tell Spin that we want to use SQLite storage, we only need to [grant SQLite permission to a component in the application’s manifest](/spin/sqlite-api-guide#granting-sqlite-database-permissions-to-components) (the `spin.toml` file). For example:

```
[component]
sqlite_databases = ["default"]
```

> The new database file (`sqlite_db.db`) is created in your application's `.spin` folder when you deploy this Spin application. The database persists across Spin application invocations and updates. The database is empty and can optionally have tables and data added using the following step.

Reviewer: ```suggestion
The new database file (`sqlite_db.db`) is created in your application's `.spin` folder when you run this Spin application. The database persists across Spin application invocations and updates. The database is empty and can optionally have tables and data added using the following step.
```

Author: Thanks @itowlson I have a new PR with this in it (and just doubled checked with Kate to use this; because Kate had a similar version/sugggestion of same)

> ## Managing You Local Spin Database Schema

Reviewer: ```suggestion
## Managing Your Local Spin Database Schema
```

Author: Thanks @itowlson 
This is in the new PR.

sqlite_databases = ["default"]
```

> The above configuration will last the lifespan of the application.

Reviewer: This is not correct - the user could change the configuration and redeploy.

Again, the original sentence aimed to communicate the lifecycle of the _database_ (and the data).

Author: Thanks; reverted in next push.

> ### Preparing Fermyon Cloud Schema

Reviewer: ```suggestion
### Create Tables and Initial Data
```

Author: Using this title in next push.

> Bootstrapping your Cloud SQLite database schema is done via the `spin cloud sqlite execute` subcommand. Your NoOps SQLite application on localhost will have already assigned a name for your database. All we need to do is obtain the name by typing the following command (in your local NoOps SQLite application's directory):

Reviewer: It's not really a "bootstrapping" process.

> Bootstrapping your Cloud SQLite database schema is done via the `spin cloud sqlite execute` subcommand. Your NoOps SQLite application on localhost will have already assigned a name for your database. All we need to do is obtain the name by typing the following command (in your local NoOps SQLite application's directory):

Reviewer: The comment about `localhost` is very misleading.  None of this is on localhost - this is _specifically_ cloud stuff - and the name is not assigned by the application but by the cloud.

Author: So when a user runs `spin up --sqlite` and creates a new table (having not deployed to the cloud at all)
 the user's installed cloud plugin is the part that has done the work of creating the database table and subsequently the `spin cloud sqlite` subcommand is in the driver's seat?

Author: Thanks, this is resolved in the next push.

> Bootstrapping your Cloud SQLite database schema is done via the `spin cloud sqlite execute` subcommand. Your NoOps SQLite application on localhost will have already assigned a name for your database. All we need to do is obtain the name by typing the following command (in your local NoOps SQLite application's directory):

Reviewer: I would cut "All we need to do is" because there are more steps after this one!  *grin*

> Bootstrapping your Cloud SQLite database schema is done via the `spin cloud sqlite execute` subcommand. Your NoOps SQLite application on localhost will have already assigned a name for your database. All we need to do is obtain the name by typing the following command (in your local NoOps SQLite application's directory):

Reviewer: This does not need to be in the application directory - it asks about global cloud state.

Author: Thanks.

> Bootstrapping your Cloud SQLite database schema is done via the `spin cloud sqlite execute` subcommand. Your NoOps SQLite application on localhost will have already assigned a name for your database. All we need to do is obtain the name by typing the following command (in your local NoOps SQLite application's directory):

Reviewer: Maybe we need to start this with more context.  This might help address the evident confusion about identifying the local and cloud databases.  E.g.:

When Fermyon Cloud creates a SQLite database, the database is empty. To be useful, you'll need to create tables, and potentially populate some of those tables with initial data.  You can do this either from within the application, or from the command line using the `spin cloud sqlite execute` command.

> When running locally, you can set up the database using `spin up --sqlite`, but this option isn't available when deploying to Fermyon Cloud.

To set up the cloud database from within your application...

To set up the cloud database from the command line...

(these latter two might need to be another level of subheading, in which case it might be worth busting this topic up a level so we're not dealing with four levels of hierarchy!)

Reviewer: I see there was a "from within your application" section in an earlier draft which was removed at Ryan's guidance, so please ignore it in the above suggestion!

<!-- @selectiveCpy -->

```bash
$ spin cloud sqlite list
Databases (1)
inspirational-pig (default)
```

Once we have this name (in this above example, we can see our database called `inspirational-pig`), we use that `inspirational-pig` as part of our `spin cloud sqlite execute` subcommand to create our schema for our Fermyon Cloud application:

```bash

> $ spin cloud sqlite execute inspirational-pig "CREATE TABLE IF NOT EXISTS todos (id INTEGER PRIMARY KEY AUTOINCREMENT,description TEXT NOT NULL,due_date DATE,starred BOOLEAN DEFAULT 0,is_completed BOOLEAN DEFAULT 0)"

Reviewer: Does `spin cloud sqlite execute` support `@file` syntax?  (Like `spin up --sqlite` does.)  If it does, it might be easier to show that than squeezing this long statement onto one line.

Author: I checked, unfortunately, it does not. Would be great to have though.

Reviewer: Sigh, yeah... the `@file` syntax is there in the cloud-plugin code (https://github.com/fermyon/cloud-plugin/blob/0ce1b4e234471a46e9a9b6688c1b24256259a597/src/commands/sqlite.rs#L101), but looks like it went in a few hours after the 0.1.2 release.

```

> As mentioned above, Fermyon Cloud only supports databases named “default” at this time. Once you add the `sqlite_databases = ["default"]` line to your local Spin application's manifest, Spin internally associates that `default` database with the random name Spin autogenerates, i.e. `inspirational-pig`. The name is not arbitrary and must be used explicitly when using the `spin cloud sqlite execute` subcommand (as shown above). This will allow you to develop your application, continually build locally, and deploy to Fermyon Cloud, knowing that the two instances (localhost and Cloud) share the same schema.

Reviewer 2: I am confused how the beginning of this paragraph leads to the ending point of the fact that the "default" name in a `spin.toml` helps you know "that the two instances (localhost and Cloud) share the same schema."

Author: Thanks so much @itowlson and @kate-goldenring 
Perhaps the approach for now (after reading the latest `spin cloud link` meeting notes etc.) is that less is more.
The docs can perhaps simply say that:
- Fermyon Cloud only supports databases named “default”
- A user can obtain the name of the database that Spin created via `spin cloud sqlite list`
- The `spin cloud sqlite execute` takes the database name as an argument.
I see some comments in the file view; I will head over and make some updates and ping in here when finished. Would be brilliant to finalise this. Thanks again for your help.

Author: Thanks this has been reworded now.

> As mentioned above, Fermyon Cloud only supports databases named “default” at this time. Once you add the `sqlite_databases = ["default"]` line to your local Spin application's manifest, Spin internally associates that `default` database with the random name Spin autogenerates, i.e. `inspirational-pig`. The name is not arbitrary and must be used explicitly when using the `spin cloud sqlite execute` subcommand (as shown above). This will allow you to develop your application, continually build locally, and deploy to Fermyon Cloud, knowing that the two instances (localhost and Cloud) share the same schema.

Reviewer: Here "named" is used in different ways - previously we had to do a list for the name and the name was "inspirational-pig", now the name is "default".  The two things in play are "what Cloud calls the actual physical database" and "how the application refers to the database" (with an implicit mapping layer between the two) - we don't have great terms for these yet though.

Author: Down to one "named"; just the original :)

> As mentioned above, Fermyon Cloud only supports databases named “default” at this time. Once you add the `sqlite_databases = ["default"]` line to your local Spin application's manifest, Spin internally associates that `default` database with the random name Spin autogenerates, i.e. `inspirational-pig`. The name is not arbitrary and must be used explicitly when using the `spin cloud sqlite execute` subcommand (as shown above). This will allow you to develop your application, continually build locally, and deploy to Fermyon Cloud, knowing that the two instances (localhost and Cloud) share the same schema.

Reviewer: This paragraph is not correct:

* "Spin internally associates that `default` database with the random name" - no, _cloud_ does all this, the way Spin maps manifest names is entirely different
* "the random name Spin autogenerates" - Spin does not generate the name
* "The name is not arbitrary" - it is totally arbitrary _grin_, I assure you nobody would choose "inspirational pig" deliberately
* "This will allow you to develop your application, continually build locally, and deploy to Fermyon Cloud, knowing that the two instances (localhost and Cloud) share the same schema." - no, the local and cloud databases are _entirely_ disconnected.  Locally, Spin gives you a SQLite database file in the `.spin` directory.  Cloud gives you a cloud-hosted database somewhere in Ruritania.

Author: Thanks, all adjusted accordingly.

### Deleting the Database

> Lastly, (warning; using the `delete` subcommand is **permanent**) you can delete your default database with the following command:

Reviewer: While we're in here, maybe bust out the "deleting is permanent" to its own emphasised paragraph, which would fix the awkward placement and punctuation.

Author: Done, thanks so much!

sqlite_databases = ["default"]
```

The above configuration will last the lifespan of the application.

### Preparing Fermyon Cloud Schema

Once you have a NoOps SQLite Storage Spin application successfully built on localhost, prepare the appropriate Fermyon Cloud schema. You need to prepare Fermyon Cloud because your local Spin application's SQLite database is independent of your Cloud application's database. You can think of your local Spin application as a development and testing environment (with its own test data) and Fermyon Cloud as the production environment (with real user data). You can bootstrap your Cloud SQLite database schema (to match your local Spin application) in a couple of different ways.

> #### Source Code

Reviewer 3: I would actually remove this section. It is possible to it this way, but I would certainly not consider it a "best practice". The "Spin Subcommand" section highlights the "best practice" way to migrate the database's schema.

Author: Thanks @rylev 
I have removed that paragraph and just retained the subcommand approach, as suggested.
